File: Ruby/day02.rb
#!/usr/bin/env ruby

require_relative 'lib/util'
#require_relative 'lib/geometry'
#require_relative 'lib/grid'

# INPUT_FILE = 'day02_test.txt'
INPUT_FILE = 'day02_challenge.txt'

input = Util.read_input("../Input/#{INPUT_FILE}")

puts 'Advent of Code 2023, Day 02: Cube Conundrum'

class CubeGame
	def initialize(defn)
		id_defn, draws_defn = defn.split(': ')
		game, id = id_defn.split(' ')
		@id = Integer(id)
		draws = draws_defn.split('; ')
		@draws = []
		draws.each do |draw|
			info = Hash.new(default=0)
			cube_counts = draw.split(', ')
			cube_counts.each do |cc|
				count, color = cc.split(' ')
				info[color] = Integer(count)
			end
			@draws.push(info)
		end
	end
	
	attr_reader :id
	
	def max_count(color)
		result = 0
		@draws.each do |draw|
			result = draw[color] if draw[color] > result
		end
		result
	end
	
	def power
		max_count('blue') * max_count('green') * max_count('red')
	end
end

def solve_part_one(games)
	cube_counts = {'red' => 12, 'green' => 13, 'blue' => 14}
	id_sum = 0
	games.each do |game|
		if game.max_count('red') <= cube_counts['red'] &&
			game.max_count('green') <= cube_counts['green'] &&
			game.max_count('blue') <= cube_counts['blue'] then
			
			id_sum += game.id
		end
	end
	puts "The sum of IDs of possible games is #{id_sum}"
end


def solve_part_two(games)
	power_sum = 0
	games.each do |game|
		power_sum += game.power
	end
	puts "The sum of power of the games is #{power_sum}"
end


def parse_games(input)
	games = []
	input.each do |line|
		game = CubeGame.new(line)
		games.push(game)
	end
	games
end

games = parse_games(input)

solve_part_one games
solve_part_two games

File: Perl/day02.pl
#!/usr/bin/env perl
use v5.40;

our $directory;
BEGIN { use Cwd; $directory = cwd; }
use lib $directory . '/lib';

use feature 'signatures';
#use Data::Printer;
use List::Util qw( max );
#use Storable 'dclone';

use AOC::Util;
#use AOC::Geometry;
#use AOC::Grid;

#my $INPUT_FILE = 'day02_test.txt';
my $INPUT_FILE = 'day02_challenge.txt';
my @input = read_input("../input/$INPUT_FILE");

say "Advent of Code 2023, Day 02: Cube Conundrum";

solve_part_one(@input);
solve_part_two(@input);

exit( 0 );

sub solve_part_one(@input) {
	my $sum_possible_ids = 0;
	for my $line (@input) {
		my $game = parse_game($line);
		if( is_game_possible($game, 12, 13, 14) ) {
			$sum_possible_ids += $game->{"id"};
		}
	}
	say "Part One: the sum of ids of possible games is: $sum_possible_ids";
}

sub solve_part_two(@input) {
	my $sum_power = 0;
	for my $line (@input) {
		my $game = parse_game($line);
		$sum_power += calc_power($game);
	}
	say "Part Two: the sum of cube power is: $sum_power";
}

sub is_game_possible($game, $r_limit, $g_limit, $b_limit) {
	my $maximums = get_maximum_observed_counts($game);
	return ($maximums->{"red"} <= $r_limit &&
			$maximums->{"green"} <= $g_limit &&
			$maximums->{"blue"} <= $b_limit);
}

sub calc_power($game) {
	my $maximums = get_maximum_observed_counts($game);
	return ($maximums->{"red"} * $maximums->{"green"} * $maximums->{"blue"});
}

sub get_maximum_observed_counts($game) {
	my @draws = @{$game->{"draws"}};
	my %maximums = ("red" => 0, "green" => 0, "blue" => 0);
	for my $draw (@draws) {
		for my $color (keys %maximums) {
			$maximums{$color} = max($maximums{$color}, $draw->{$color});
		}
	}
	return \%maximums;
}

sub parse_game($line) {
	my %game = ();
	my ($game_info, $draws_str) = split(": ", $line);
	$game_info =~ m/(\d+)/;
	$game{"id"} = $1;

	my @draws = ();
	my @each_draw_string = split("; ", $draws_str);
	for my $draw_string (@each_draw_string) {
		my %draw = ("red" => 0, "green" => 0, "blue" => 0);
		while ( $draw_string =~ m/(\d+) (red|green|blue)/g ) {
			$draw{$2} = $1;
		}
		push(@draws, \%draw);
	}
	$game{"draws"} = \@draws;
	return \%game;
}